src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java

```diff
@@ -0,0 +1,78 @@
+import org.jabref.logic.util.io.AutoLinkPreferences;
+import org.jabref.logic.util.io.FileFinder;
+import org.jabref.logic.util.io.FileFinders;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+import org.jabref.model.metadata.FileDirectoryPreferences;
+import org.jabref.model.util.FileHelper;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class AutoSetFileLinksUtil {
+
+    private static final Log LOGGER = LogFactory.getLog(AutoSetLinks.class);
+
+    public Map<BibEntry, LinkedFile> findassociatedNotLinkedFiles(BibEntry entry, BibDatabaseContext databaseContext, FileDirectoryPreferences fileDirPrefs, AutoLinkPreferences autoLinkPrefs) {
+        return findassociatedNotLinkedFiles(Arrays.asList(entry), databaseContext, fileDirPrefs, autoLinkPrefs);
+    }
+
+    public Map<BibEntry, LinkedFile> findassociatedNotLinkedFiles(List<BibEntry> entries, BibDatabaseContext databaseContext, FileDirectoryPreferences fileDirPrefs, AutoLinkPreferences autoLinkPrefs) {
+        Map<BibEntry, LinkedFile> linkedFiles = new HashMap<>();
+
+        List<Path> dirs = databaseContext.getFileDirectoriesAsPaths(fileDirPrefs);
+        List<String> extensions = ExternalFileTypes.getInstance().getExternalFileTypeSelection().stream().map(ExternalFileType::getExtension).collect(Collectors.toList());
+
+        // Run the search operation:
+        FileFinder fileFinder = FileFinders.constructFromConfiguration(autoLinkPrefs);
+        Map<BibEntry, List<Path>> result = fileFinder.findAssociatedFiles(entries, dirs, extensions);
+
+        // Iterate over the entries:
+        for (Entry<BibEntry, List<Path>> entryFilePair : result.entrySet()) {
+
+            for (Path foundFile : entryFilePair.getValue()) {
+                boolean existingSameFile = entryFilePair.getKey().getFiles().stream()
+                        .map(file -> file.findIn(dirs))
+                        .anyMatch(file -> {
+                            try {
+                                return file.isPresent() && Files.isSameFile(file.get(), foundFile);
+                            } catch (IOException e) {
+                                LOGGER.error("Problem with isSameFile", e);
+                            }
+                            return false;
+                        });
+                if (!existingSameFile) {
+
+                    Optional<ExternalFileType> type = FileHelper.getFileExtension(foundFile)
+                            .map(ExternalFileTypes.getInstance()::getExternalFileTypeByExt)
+                            .orElse(Optional.of(new UnknownExternalFileType("")));
+
+                    String strType = type.isPresent() ? type.get().getName() : "";
+
+                    LinkedFile linkedFile = new LinkedFile("", foundFile.toString(), strType);
+                    linkedFiles.put(entryFilePair.getKey(), linkedFile);
```

Comment:
If I see it correctly, we are only returning one file per entry although the file finder may find more than 1, right?

Personally, I would restructure the code a bit more: instead of calling `findassociatedNotLinkedFiles(List<BibEntry> entries,...` in the method `findassociatedNotLinkedFiles(BibEntry entry,...`, I would iterate over all bib entries and call the method for on entry, i.e.
````java
findassociatedNotLinkedFiles(List<BibEntry> entries) {
     for(entry in entries) {
           append(findassociatedNotLinkedFiles(entry));
      }
}
````
Or is there is good reason to not do this?

Reply:
I changed the code so that multiple files will be added, except if there already exists the same file. That's why I return a map now
In the lines above I iterate over all pairs of entries and files

Reply:
Sorry, I don't really understand what you mean. Since you return a map, there is by definition only one returned file per entry (since the key in the map is unique). What you need is a multimap (we use a library that provides these...no idea which...guava?)

Reply:
Ah yes, did not have this in mind. Will think about another solution

Reply:
I think the code gets considerably more easy if you just do the check for one entry (and not a collection) and return a list of to-be-added links. Then the caller can iterate over it.